Removal of redundant functions

JulStraus opened the pull request:
The rework regarding both [storage investments](https://github.com/EnergyModelsX/EnergyModelsInvestments.jl/pull/21) and [investment modes](https://github.com/EnergyModelsX/EnergyModelsInvestments.jl/pull/23) allows for a significant reduction of functions through utilization of the newly introduced function `add_investment_constraints`. In addition, it furthermore allows making `EnergyModelsInvestments` independent of `EnergyModelsBase` through moving the relevant functions to `EnergyModelsBase`.

This rework is introduced in this PR. It closes [Issue 20](https://github.com/EnergyModelsX/EnergyModelsInvestments.jl/issues/20).

Due to the rework, I additionally changed the folder structure with respect to the definition of the individually introduced types.

hellemo commented on src/utils.jl:
```diff
@@ -1,489 +1,195 @@
+- `years:`: the remaining years for calculating the discounted value. The years are
+  depending on the considered [`LifetimeMode`](@ref), using `remaining(t_inv, 𝒯)` for
+  [`StudyLife`](@ref) and `duration(t_inv)` for [`PeriodLife`](@ref).
+- `lifetime`: the lifetime of the ttype.
```
```suggestion
- `lifetime`: the lifetime of the type.
```
Or perhaps rather `element` (see above)

hellemo commented on src/model.jl:
```diff
@@ -134,61 +133,61 @@ function EMB.variables_capex(m, 𝒩, 𝒯, 𝒫, modeltype::AbstractInvestmentM
-* bounds
-* binary for BinaryInvestment
-* link capacity variables
+When the modeltype is an investment model, the function introduces the related constraints
```
```suggestion
Add capacity expansion constraints to model `m` when `modeltype` is investment model. 
```

JulStraus replied:
I am a bit opposed to the rephrasing here. The approach of using `when` highlights with change has to happen so that we use the function.

hellemo commented on src/model.jl:
```diff
@@ -134,61 +133,61 @@ function EMB.variables_capex(m, 𝒩, 𝒯, 𝒫, modeltype::AbstractInvestmentM
-* binary for BinaryInvestment
-* link capacity variables
+When the modeltype is an investment model, the function introduces the related constraints
+for the capacity expansion. The investment mode and lifetime mode are used for adding
```
```suggestion
The investment mode and lifetime mode are used for adding
```

hellemo replied:
NP, I think I missed the context and worded it like it was the first sentence, but since it's not we can keep as is.

hellemo commented on src/model.jl:
```diff
@@ -134,61 +133,61 @@ function EMB.variables_capex(m, 𝒩, 𝒯, 𝒫, modeltype::AbstractInvestmentM
+constraints.
 
+The default function only accepts nodes with [`SingleInvData`](@ref). If you have several
+capacities for investments, you have to dispatch specifically on the function. This is
```
```suggestion
capacities for investments, you have to dispatch specifically on the node type. This is
```
?

hellemo commented on src/model.jl:
```diff
@@ -214,14 +213,53 @@ function EMB.constraints_capacity_installed(m, n::Storage, 𝒯::TimeStructure,
+Core function for introducing constraints on the investments. The constraints include
+introducing bounds on the available capacities as well as the calculation of the CAPEX.
+
+The function calls two in addition subroutines, [`set_capacity_installation`](@ref) and
```
```suggestion
The function calls two additional subroutines, [`set_capacity_installation`](@ref) and
```

hellemo commented on src/model.jl:
```diff
@@ -214,14 +213,53 @@ function EMB.constraints_capacity_installed(m, n::Storage, 𝒯::TimeStructure,
+
+The function calls two in addition subroutines, [`set_capacity_installation`](@ref) and
+[`set_capacity_cost`](@ref) which are used for introducing bounds on the investment
+variables and calculating the CAPEX contribution of each investments. The utilizaiton of
```
```suggestion
variables and calculating the CAPEX contribution of each investments. The utilization of
```

hellemo reviewed (commented):
This looks good. I have added comments mostly to address minor issues and suggestions for rephrasing docs and fix typos.

One topic for discussion is what name to use instead of `node`/`n`. I assume `type` was used to make it more general (could apply also to edges/links). I think `type` is confusing here due to the common discussion on dispatch on different types (in the PL sense), so I suggest we call it something else. My first suggestion would be `element`, but open for better ideas.

JulStraus commented:
> This looks good. I have added comments mostly to address minor issues and suggestions for rephrasing docs and fix typos.
> 
> One topic for discussion is what name to use instead of `node`/`n`. I assume `type` was used to make it more general (could apply also to edges/links). I think `type` is confusing here due to the common discussion on dispatch on different types (in the PL sense), so I suggest we call it something else. My first suggestion would be `element`, but open for better ideas.

I updated it using `element`. Generally speaking, I think that improves the description.
